### PR TITLE
Fix DHCP option order not being preserved

### LINKED ISSUE #75: Encoder produces different results
```rust
pub fn create_pxe_find_packet(chaddr: [u8; 6]) -> Vec<u8> {
    let message_type = v4::MessageType::Discover;
    let params = v4::DhcpOption::ParameterRequestList(vec![
        v4::OptionCode::SubnetMask,
        v4::OptionCode::Router,
        v4::OptionCode::DomainNameServer,
        v4::OptionCode::TFTPServerName,
        v4::OptionCode::BootfileName,
    ]);

    let mut msg = v4::Message::default();

    msg.set_flags(v4::Flags::default().set_broadcast())
        .set_chaddr(&chaddr);

    msg.opts_mut().insert(v4::DhcpOption::MessageType(v4::MessageType::Discover));

    msg.opts_mut().insert(params);

    msg.opts_mut().insert(v4::DhcpOption::End);

    let mut buf = Vec::new();
    let mut e = Encoder::new(&mut buf);
    msg.encode(&mut e).unwrap();

    buf
}
```

### PR DESCRIPTION
**Pull request by f3rn0s**
Fixes #75 by introducing a data-structure that will preserve insertion order.

### CODE REVIEW

**Comment by leshow** [comment]
To be clear, removing the manual insert of `End` fixes your issue where the encoder did not produce a valid message right? It is just not going to be byte-by-byte the same.

**Comment by f3rn0s** [comment]
Sorry for the delay on this. I'm unsure why this would break backwards compatibility. Users don't particular interact with the internal Options object.

It's giving 100% test passing on my machine, it looked like last time was a serde issue, so hopefully by explicitly enabling the `serde` feature of IndexMap this won't be a problem.

Also imo I think that a protocol encoding library should be byte-by-byte accurate between runs.

**Comment by stappersg** [comment]
On Wed, Oct 09, 2024 at 09:58:54PM -0700, f3rn0s wrote:
> 
> Also imo I think that a protocol encoding library
> should be byte-by-byte accurate between runs.
> 

I think that it is good thing if server replies to options in the same
order as original requested by the client. ( In other words: Do not
assume that both ends can deal with a reshuffled option order. )


Groeten
Geert Stappers
-- 
Silence is hard to parse

**Comment by leshow** [comment]
@f3rn0s the iterator type makes it into the public API in `IntoIter`, that's the spot I'm concerned about the breaking change

Have either of you seen anything in RFC's which implied an option order?

I lean more towards btreemap because it's in the standard lib, requires no dependencies, and satisfies your concern of producing the same byte-by-byte output.

**Comment by f3rn0s** [comment]
@leshow I opted for IndexMap since that preserves insertion order over just a consistent, but not necessarily visible/modifiable, order.

There was no reference to ordering in RFC 2131 or 2132, but it would be nice to have granular control in case we came across a server or client implementation that has weird requirements.

**Comment by leshow** [comment]
I went for btree since it doesn't require a dependency
